twa

```diff
@@ -391,8 +392,78 @@ function stage_4_scm_repos {
 # * Fetch the server's cert using `openssl s_client`, check it for known bad suites
 #   * Alternatively, just `nmap --script ssl-enum-ciphers -p 443 $domain`
 
+function stage_5_CAA {
+  verbose "Stage 5: Certification Authority Authorization"
+  verbose "Which Certificate Authorities (CAs) are allowed to issue certificates for the domain"
+    code=`dig +short caa $domain`
+    number_of_lines=`echo $code|wc -l`
+    if [ $number_of_lines -gt 1 ];then
+      INFO "multiple CAAs specified, everyone join the party"
+      #Cloudflare currently set the following set of certs as CAA
+      if [ echo $code|grep -c 'issue “comodoca.com”' -gt 0 ];then
+        if [ echo $code|grep -c 'issue “digicert.com”' -gt 0 ];then
+          if [ echo $code|grep -c 'issue “globalsign.com”' -gt 0 ];then
+            if [ echo $code|grep -c 'issuewild “comodoca.com”' -gt 0 ];then
+              if [ echo $code|grep -c 'issuewild “digicert.com”' -gt 0 ];then
+                if [ echo $code|grep -c 'issuewild “globalsign.com”' -gt 0 ];then
+                  INFO "CAA is probably provided by Cloudflare"
+                fi
+              fi 
+            fi
+          fi
+        fi
+      fi
+    fi
+
+
+    echo $code| while read line ; do
+      flag=`echo $code |cut -d' ' -f1| sed -e 's/^"//' -e 's/"$//'`
+      tag=`echo $code |cut -d' ' -f2`
+      value=`echo $code|cut -d' ' -f3| sed -e 's/^"//' -e 's/"$//'`
+      #list does not match returned names at this stage
+      #if [ `grep -c $value certs/untrustcert.txt` -ne 0 ];then    
+      #  FAIL "CA is listed as untrusted!"
+      #elif [ `grep -c $value certs/trustcert.txt` -ne 0 ];then        
+      #   PASS "CA is listed as trusted"
+      #else
+      #   UNK "Got an unknown CA:"$code
+      #fi
+
+      if [[ -z "$code" ]]; then
+         FAIL "No CAA record or doesn't specify an issuer at all. Potentially vunerable to have CA changed/MTM"
+      elif [[ ${code} == ";" ]]; then
+         MEH "CAA disallowed all CAA issuers"
+      elif [ `echo $tag| grep -c "issue"` -ne 1 ]; then
+         FAIL "CAA incorectly specified"
+      elif [ ${flag} -eq 0 ] && [ $tag == "issue" ]; then
+         PASS "CAA issuer:"$value
+      else
+         UNK "Got a weird response code:"$code
+      fi
+    done
+
+    nmap $domain|grep open|cut -d/ -f1 | while read line ; do
+      INFO "port open:"$line
+      if [ $line -ne 80 ];then
+        if [ $(nmap --script ssl-cert,ssl-enum-ciphers -p $line $domain|grep -c weak) -ne 0 ];then
+          FAIL "Weak ciphers detected"
+        else
+          PASS "ciphers ok"
+        fi
+      fi
+    done
+
+    owner=$(whois $domain|grep Registrant:)
```

Comment:
This is also out of scope.

I'm also not convinced of its value: most people *do* want to hide their registrant information, for privacy reasons. So if we were to have this test in the future, we'd probably flip the `PASS` and `FAIL` around.

Reply:
If the site belongs to a company or corporation, there is no need to hide that information. If the site is someone pretending to be a business, it is nice to know they have hidden it for some reason.

Reply:
I agree that those are interesting cases to consider, but I think they're asking subtly different questions than `twa` is meant to ask: both are asking questions about the entity or identity behind the domain, while `twa` is just meant to ask whether the domain has good security defaults.

Importantly, that means that `twa` should give the same score to `realcompany.com` and `evilphisher.com` if they provide the same security features, even if `evilphisher.com` is malicious.